ldap/servers/slapd/csngen.c

```diff
@@ -597,94 +557,93 @@ _csngen_cmp_callbacks(const void *el1, const void *el2)
+        return CSN_TIME_ERROR;
+    }
+    cur_time = now.tv_sec;
+    time_diff = cur_time - gen->state.sampled_time;
 
+    /* check if the time should be adjusted */
     if (time_diff == 0) {
         /* This is a no op - _csngen_adjust_local_time should never be called
            in this case, because there is nothing to adjust - but just return
            here to protect ourselves
         */
         return CSN_SUCCESS;
-    } else if (time_diff > 0) {
-        time_t ts_before = CSN_CALC_TSTAMP(gen);
-        time_t ts_after = 0;
-        if (slapi_is_loglevel_set(SLAPI_LOG_REPL)) {
-            time_t new_time = CSN_CALC_TSTAMP(gen);
-            slapi_log_err(SLAPI_LOG_REPL, "_csngen_adjust_local_time",
-                          "gen state before %08lx%04x:%ld:%ld:%ld\n",
-                          new_time, gen->state.seq_num,
-                          gen->state.sampled_time,
-                          gen->state.local_offset,
-                          gen->state.remote_offset);
-        }
-
-        gen->state.sampled_time = cur_time;
-        if (time_diff > gen->state.local_offset)
-            gen->state.local_offset = 0;
-        else
-            gen->state.local_offset = gen->state.local_offset - time_diff;
-
-        /* only reset the seq_num if the new timestamp part of the CSN
-           is going to be greater than the old one - if they are the
-           same after the above adjustment (which can happen if
-           csngen_adjust_time has to store the offset in the
-           local_offset field) we must not allow the CSN to regress or
-           generate duplicate numbers */
-        ts_after = CSN_CALC_TSTAMP(gen);
-        if (ts_after > ts_before) {
-            gen->state.seq_num = 0; /* only reset if new time > old time */
-        }
-
-        if (slapi_is_loglevel_set(SLAPI_LOG_REPL)) {
-            time_t new_time = CSN_CALC_TSTAMP(gen);
-            slapi_log_err(SLAPI_LOG_REPL, "_csngen_adjust_local_time",
-                          "gen state after %08lx%04x:%ld:%ld:%ld\n",
-                          new_time, gen->state.seq_num,
-                          gen->state.sampled_time,
-                          gen->state.local_offset,
-                          gen->state.remote_offset);
-        }
-        return CSN_SUCCESS;
-    } else /* time was turned back */
-    {
-        if (slapi_is_loglevel_set(SLAPI_LOG_REPL)) {
-            time_t new_time = CSN_CALC_TSTAMP(gen);
-            slapi_log_err(SLAPI_LOG_REPL, "_csngen_adjust_local_time",
-                          "gen state back before %08lx%04x:%ld:%ld:%ld\n",
-                          new_time, gen->state.seq_num,
-                          gen->state.sampled_time,
-                          gen->state.local_offset,
-                          gen->state.remote_offset);
-        }
+    }
+    if (labs(time_diff) > _SEC_PER_DAY) {
+        /* We had a jump larger than a day */
+        slapi_log_err(SLAPI_LOG_INFO, "csngen_new_csn",
```

Comment:
If we have a topology without any update for a day (direct or replicated). is _csngen_adjust_local_time called more frequently than a day ? I wonder if this log could occur on an update after a long quiet period. In such case there is not an alarming big jump but just a long period of quiet.

Reply:
It will log the message, I had originally added this logging line, and it will report an INFO message when it sees large jumps, even idle jumps.  I had proposed we could bump the threshold to something larger than a day.  30 days or something like that just to keep the logs quieter.